File: Контест 2/e.cpp
#include <iostream>
#include <vector>

bool Equal(std::pair<int, int> left, std::pair<int, int> right) {
  return left.first == right.first;
}

bool SortCondition(std::pair<int, int> left, std::pair<int, int> right) {
  return left.first <= right.first;
}

std::vector<std::pair<int, int>> Merge(std::vector<std::pair<int, int>>& values,
                                       std::vector<int>& answer, size_t left,
                                       size_t right) {
  size_t medium = (left + right) / 2;
  size_t left_part_index = left;
  size_t right_part_index = medium;
  std::vector<std::pair<int, int>> sorted_array(0);
  while (left_part_index < medium || right_part_index < right) {
    if (left_part_index < medium && right_part_index < right) {
      if (SortCondition(values[left_part_index], values[right_part_index])) {
        answer[values[left_part_index].second] += (right_part_index - medium);
        sorted_array.push_back(values[left_part_index++]);
      } else {
        sorted_array.push_back(values[right_part_index++]);
      }
    } else if (left_part_index < medium) {
      answer[values[left_part_index].second] += (right_part_index - medium);
      sorted_array.push_back(values[left_part_index++]);
    } else {
      sorted_array.push_back(values[right_part_index++]);
    }
  }
  return sorted_array;
}

void MergeSort(std::vector<std::pair<int, int>>& values,
               std::vector<int>& answers, size_t left = 0, size_t right = 0) {
  if (right == 0) {
    right = values.size();
  }
  if (right - left == 2) {
    if (!SortCondition(values[left], values[right - 1])) {
      ++answers[values[left].second];
      std::swap(values[left], values[right - 1]);
    }
  } else if (right - left > 2) {
    size_t medium = (left + right) / 2;
    MergeSort(values, answers, left, medium);
    MergeSort(values, answers, medium, right);
    auto sorted_array = Merge(values, answers, left, right);
    for (size_t i = left; i < right; ++i) {
      values[i] = sorted_array[i - left];
    }
  }
}

int main() {
  int amount_of_elements;
  std::cin >> amount_of_elements;
  std::vector<std::pair<int, int>> values(amount_of_elements);
  int index = 0;
  for (auto& element : values) {
    std::cin >> element.first;
    element.second = index;
    ++index;
  }
  std::vector<int> answers(values.size());
  MergeSort(values, answers);
  for (auto element : answers) {
    std::cout << element << " ";
  }
}

File: Контест 2/e.go
package main

import "fmt"

type Pair[T any] struct {
	First  T
	Second T
}

func Equal(a, b Pair[int]) bool {
	return a.First == b.First
}

func SortCondition(a, b Pair[int]) bool {
	return a.First <= b.First
}

func Merge(values *[]Pair[int], answers *[]int, left, right int) []Pair[int] {
	medium := (left + right) / 2
	left_part_index := left
	right_part_index := medium
	sorted_array := make([]Pair[int], right-left)
	for left_part_index < medium || right_part_index < right {
		if left_part_index < medium && right_part_index < right {
			if SortCondition((*values)[left_part_index], (*values)[right_part_index]) {
				(*answers)[(*values)[left_part_index].Second] += right_part_index - medium
				sorted_array = append(sorted_array, (*values)[left_part_index])
				left_part_index++
			} else {
				sorted_array = append(sorted_array, (*values)[right_part_index])
				right_part_index++
			}
		} else if left_part_index < medium {
			(*answers)[(*values)[left_part_index].Second] += right_part_index - medium
			sorted_array = append(sorted_array, (*values)[left_part_index])
			left_part_index++
		} else {
			sorted_array = append(sorted_array, (*values)[right_part_index])
			right_part_index++
		}
	}
	return sorted_array
}

func MergeSort(values *[]Pair[int], answers *[]int, left, right int) {
	if right-left == 2 {
		if !SortCondition((*values)[left], (*values)[right-1]) {
			(*answers)[(*values)[left].Second]++
			(*values)[left], (*values)[right-1] = (*values)[right-1], (*values)[left]
		}
	} else if right-left > 2 {
		MergeSort(values, answers, left, (left+right)/2)
		MergeSort(values, answers, (left+right)/2, right)
		sorted_array := Merge(values, answers, left, right)
		for i := left; i < right; i++ {
			(*values)[i] = sorted_array[i-left]
		}
	}
}

func main() {
	var amount_of_elements int
	fmt.Scan(&amount_of_elements)
	var values = make([]Pair[int], amount_of_elements)
	index := 0
	for e := range values {
		fmt.Scan(&values[e].First)
		values[e].Second = index
		index++
	}
	var answers = make([]int, amount_of_elements)
	MergeSort(&values, &answers, 0, amount_of_elements)
	for e := range answers {
		fmt.Print(answers[e], " ")
	}
}
